File: vol2/vol2-scala-examples/src/main/scala/com/heatonresearch/aifh/evolutionary/score/parallel/ParallelScoreTask.scala
/*
 * Artificial Intelligence for Humans
 * Volume 2: Nature Inspired Algorithms
 * Java Version
 * http://www.aifh.org
 * http://www.jeffheaton.com
 *
 * Code repository:
 * https://github.com/jeffheaton/aifh
 *
 * Copyright 2014 by Jeff Heaton
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * For more information on Heaton Research copyrights, licenses
 * and trademarks visit:
 * http://www.heatonresearch.com/copyright
 */
package com.heatonresearch.aifh.evolutionary.score.parallel

import com.heatonresearch.aifh.AIFHError
import com.heatonresearch.aifh.evolutionary.genome.Genome
import com.heatonresearch.aifh.evolutionary.score.AdjustScore
import com.heatonresearch.aifh.evolutionary.train.basic.BasicEA
import com.heatonresearch.aifh.learning.MLMethod
import com.heatonresearch.aifh.learning.score.ScoreFunction

/**
 * An individual threadable task for the parallel score calculation.
 * @param genome   The genome to calculate the score for.
 * @param owner The owner.
 */
class ParallelScoreTask(val genome: Genome,val owner: ParallelScore) extends Runnable {

  /**
   * The score function.
   */
  private val scoreFunction: ScoreFunction = owner.scoreFunction
  /**
   * The score adjusters.
   */
  private val adjusters: List[AdjustScore] = owner.adjusters

  /**
   * Perform the task.
   */
  def run(): Unit = {
    val phenotype = owner.codec.decode(genome)
    if (phenotype != null) {
      val score =
        try {
          scoreFunction.calculateScore(phenotype)
        } catch {
          case e: AIFHError =>
            Double.NaN
        }
      genome.score = score
      genome.adjustedScore = score
      BasicEA.calculateScoreAdjustment(genome, adjusters)
    }
  }
}

File: vol2/vol2-csharp-examples/AIFH-Vol2-Core/Core/Evolutionary/Score/Multi/ParallelScoreTask.cs
// Artificial Intelligence for Humans
// Volume 2: Nature-Inspired Algorithms
// C# Version
// http://www.aifh.org
// http://www.jeffheaton.com
//
// Code repository:
// https://github.com/jeffheaton/aifh
//
// Copyright 2014 by Jeff Heaton
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// For more information on Heaton Research copyrights, licenses
// and trademarks visit:
// http://www.heatonresearch.com/copyright
//
using AIFH_Vol2.Core.Evolutionary.Genome;
using AIFH_Vol2.Core.Evolutionary.Train;
using AIFH_Vol2.Core.Learning;
using AIFH_Vol2.Core.Learning.Score;
//
// Encog(tm) Core v3.2 - .Net Version
// http://www.heatonresearch.com/encog/
//
// Copyright 2008-2014 Heaton Research, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//   
// For more information on Heaton Research copyrights, licenses 
// and trademarks visit:
// http://www.heatonresearch.com/copyright
//
using System;
using System.Collections.Generic;

namespace AIFH_Vol2.Core.Evolutionary.Score.Multi
{
    public class ParallelScoreTask
    {
        /// <summary>
        ///     The score adjusters.
        /// </summary>
        private readonly IList<IAdjustScore> adjusters;

        /// <summary>
        ///     The genome to calculate the score for.
        /// </summary>
        private readonly IGenome genome;

        /// <summary>
        ///     The owners.
        /// </summary>
        private readonly ParallelScore owner;

        /// <summary>
        ///     The score function.
        /// </summary>
        private readonly IScoreFunction scoreFunction;

        /// <summary>
        ///     Construct the parallel task.
        /// </summary>
        /// <param name="genome">The genome.</param>
        /// <param name="theOwner">The owner.</param>
        public ParallelScoreTask(IGenome genome, ParallelScore theOwner)
        {
            owner = theOwner;
            this.genome = genome;
            scoreFunction = theOwner.ScoreFunction;
            adjusters = theOwner.Adjusters;
        }

        /// <summary>
        ///     Perform the task.
        /// </summary>
        public void PerformTask()
        {
            IMLMethod phenotype = owner.CODEC.Decode(genome);
            if (phenotype != null)
            {
                double score;
                try
                {
                    score = scoreFunction.CalculateScore(phenotype);
                }
                catch (AIFHError e)
                {
                    score = Double.NaN;
                }
                genome.Score = score;
                genome.AdjustedScore = score;
                BasicEA.CalculateScoreAdjustment(genome, adjusters);
            }
        }
    }
}
